src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -20,83 +15,126 @@ public partial class SocketAsyncEventArgs : EventArgs, IDisposable
         private byte[] _controlBuffer;
         private GCHandle _controlBufferGCHandle;
-        private IntPtr _ptrControlBuffer;
-        private WSABuffer[] _wsaRecvMsgWSABufferArray;
-        private GCHandle _wsaRecvMsgWSABufferArrayGCHandle;
-        private IntPtr _ptrWSARecvMsgWSABufferArray;
-
-        // Internal buffer for AcceptEx when Buffer not supplied.
-        private IntPtr _ptrAcceptBuffer;
 
-        // Internal SocketAddress buffer
-        private GCHandle _socketAddressGCHandle;
-        private Internals.SocketAddress _pinnedSocketAddress;
-        private IntPtr _ptrSocketAddressBuffer;
-        private IntPtr _ptrSocketAddressBufferSize;
+        // SocketAddress GCHandle
+        private GCHandle _socketAddressBufferGCHandle;
 
         // SendPacketsElements property variables.
         private SendPacketsElement[] _sendPacketsElementsInternal;
-        private Interop.Winsock.TransmitPacketsElement[] _sendPacketsDescriptor;
         private int _sendPacketsElementsFileCount;
         private int _sendPacketsElementsBufferCount;
 
         // Internal variables for SendPackets
         private FileStream[] _sendPacketsFileStreams;
         private SafeHandle[] _sendPacketsFileHandles;
-        private IntPtr _ptrSendPacketsDescriptor;
 
         // Overlapped object related variables.
         private SafeNativeOverlapped _ptrNativeOverlapped;
         private PreAllocatedOverlapped _preAllocatedOverlapped;
-        private object[] _objectsToPin;
+
         private enum PinState
         {
-            None = 0,
-            NoBuffer,
+            NoBuffer = 0,
             SingleAcceptBuffer,
             SingleBuffer,
             MultipleBuffer,
             SendPackets
         }
         private PinState _pinState;
-        private byte[] _pinnedAcceptBuffer;
-        private byte[] _pinnedSingleBuffer;
-        private int _pinnedSingleBufferOffset;
-        private int _pinnedSingleBufferCount;
 
         internal int? SendPacketsDescriptorCount
         {
             get
             {
-                return _sendPacketsDescriptor == null ? null : (int?)_sendPacketsDescriptor.Length;
+                return _sendPacketsElementsInternal == null ? null : (int?)_sendPacketsElementsFileCount + _sendPacketsElementsBufferCount;
             }
         }
 
         private void InitializeInternals()
         {
             // Zero tells TransmitPackets to select a default send size.
             _sendPacketsSendSize = 0;
+
+            AllocateOverlapped();
         }
 
-        private void FreeInternals(bool calledFromFinalizer)
+        private void FreeInternals()
         {
-            // Free native overlapped data.
-            FreeOverlapped(calledFromFinalizer);
+            ClearPinState();
+
+            // Free the overlapped object.
+            if (_ptrNativeOverlapped != null && !_ptrNativeOverlapped.IsInvalid)
+            {
+                _ptrNativeOverlapped.Dispose();
```

Comment:
We probably still want to Dispose of it even if it's invalid, no?  Otherwise it may still get finalized (only to then see that it's invalid and there's nothing to do).